src/canister/individual_user_template/src/api/cdao/mod.rs

```diff
@@ -0,0 +1,266 @@
+
+async fn install_canister_wasm(
+    wasm: Vec<u8>,
+    arg: Vec<u8>,
+    canister_id: PrincipalId,
+) -> Result<(), CdaoDeployError> {
+    let install_arg = InstallCodeArgument {
+        mode: CanisterInstallMode::Install,
+        canister_id: canister_id.0,
+        wasm_module: wasm,
+        arg,
+    };
+    install_code(install_arg).await?;
+    Ok(())
+}
+
+async fn update_controllers(
+    canister_id: PrincipalId,
+    controllers: Vec<Principal>,
+) -> Result<(), CdaoDeployError> {
+    update_settings(UpdateSettingsArgument {
+        canister_id: canister_id.0,
+        settings: CanisterSettings {
+            controllers: Some(controllers),
+            ..Default::default()
+        },
+    })
+    .await?;
+    Ok(())
+}
+
+#[query]
+async fn deployed_cdao_canisters() -> Vec<DeployedCdaoCanisters> {
+    CANISTER_DATA.with(|cdata| {
+        cdata
+            .borrow()
+            .cdao_canisters
+            .clone()
+    })
+}
+
+#[update]
+async fn deploy_cdao_sns(
+    init_payload: SnsInitPayload,
+) -> Result<DeployedCdaoCanisters, CdaoDeployError> { 
+    // * access control
+    let current_caller = ic_cdk::caller();
+    let my_principal_id = CANISTER_DATA
+        .with(|canister_data_ref_cell| canister_data_ref_cell.borrow().profile.principal_id);
+    if my_principal_id != Some(current_caller) {
+        return Err(CdaoDeployError::Unauthenticated);
+    };
+
+    let (registered, limit_hit) = CANISTER_DATA.with(|cdata| {
+        let cdata = cdata.borrow();
+        let registered = matches!(cdata.session_type, Some(SessionType::RegisteredSession));
+        (registered, cdata.cdao_canisters.len() == CDAO_TOKEN_LIMIT)
+    });
+    /*if !registered {
+        return Err(CdaoDeployError::Unregistered);
+    }*/
+    if limit_hit {
+        return Err(CdaoDeployError::TokenLimit(CDAO_TOKEN_LIMIT));
+    }
+
+    let creation_arg = CreateCanisterArgument {
+        settings: Some(CanisterSettings {
+            controllers: Some(vec![ic_cdk::id()]),
+            ..Default::default()
+        }),
+    };
+
+    let canisters: Vec<_> = (0..5)
+        .map(|_| create_empty_canister(creation_arg.clone()))
+        .collect::<FuturesOrdered<_>>()
+        .try_collect()
+        .await?;
+    let governance = canisters[0];
+    let ledger = canisters[1];
+    let root = canisters[2];
+    let swap = canisters[3];
+    let index = canisters[4];
+
+    let sns_canisters = SnsCanisterIds {
+        governance,
+        ledger,
+        root,
+        swap,
+        index,
+    };
+    let mut payloads = init_payload
+        .build_canister_payloads(&sns_canisters, None, true)
+        .map_err(CdaoDeployError::InvalidInitPayload)?;
+    let time_seconds = ic_cdk::api::time() / 1_000_000_000;
+    payloads.swap.swap_start_timestamp_seconds = Some(time_seconds);
+    payloads.swap.swap_due_timestamp_seconds = Some(time_seconds + 300);
+    payloads.swap.icp_ledger_canister_id = "ryjl3-tyaaa-aaaaa-aaaba-cai".to_string();
```

Comment:
is it fine to set it to this?

Reply:
- `ryjl3-tyaaa-aaaaa-aaaba-cai` this is anyways the mainnet canister, we are setting this just so local testing is easier (for eg there's a setup_icp_ledger.sh to deploy the icp ledger on local)
- the timeout of 5 mins is the initial swap time. 5 min is the max, can also set lower. i was considering if i should make it a separate input param itself to `deploy_cado_sns`

Reply:
i added swap_time as a parameter directly